File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        res = ListNode(0)
        curr = res
        carry = 0
        while l1 and l2:
            curr_d = l1.val + l2.val + carry
            if curr_d < 10:
                curr.next = ListNode(curr_d)
                carry = 0
            else:
                digit = curr_d % 10
                curr.next = ListNode(digit)
                carry = 1
            curr = curr.next
            l1 = l1.next
            l2 = l2.next
        
        while l1:
            curr_d = l1.val + carry
            if curr_d < 10:
                curr.next = ListNode(curr_d)
                carry = 0
            else:
                digit = curr_d % 10
                curr.next = ListNode(digit)
                carry = 1
            curr = curr.next
            l1 = l1.next
        
        while l2:
            curr_d = l2.val + carry
            if curr_d < 10:
                curr.next = ListNode(curr_d)
                carry = 0
            else:
                digit = curr_d % 10
                curr.next = ListNode(digit)
                carry = 1
            curr = curr.next
            l2 = l2.next
        
        if carry != 0:
            curr.next = ListNode(carry)


        return res.next

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length == 0) return null;
        // if(lists.length < 2) return lists[0];

        for(int i=1; i<lists.length; i++){
            if(lists[i] == null || lists[i-1] == null) continue;
            ListNode curr = mergeTwoLists(lists[i], lists[i-1]);
            lists[i] = curr;
        }
        return lists[lists.length-1];
    }

    public ListNode mergeTwoLists(ListNode l1, ListNode l2){
        ListNode dummy = new ListNode(0);
        ListNode res = dummy;

        while(l1 != null && l2 != null){
            if(l1.val <= l2.val){
                res.next = l1;
                l1 = l1.next;
            }else{
                res.next = l2;
                l2 = l2.next;
            }
            res = res.next;
        }
        if(l1 != null){
            res.next = l1;
        }else{
            res.next = l2;
        }
        return dummy.next;
    }
}
